felix/bpf-gpl/qos.h

```diff
@@ -111,4 +112,44 @@ static CALI_BPF_INLINE int enforce_packet_rate_qos(struct cali_tc_ctx *ctx)
 	return TC_ACT_SHOT;
 }
 
+static CALI_BPF_INLINE int set_dscp(struct cali_tc_ctx *ctx)
+{
+#if (CALI_F_FROM_WEP || CALI_F_TO_HEP)
+	// TODO (mazdak): set DSCP only if traffic is leaving cluster
+	if (EGRESS_DSCP < 0) {
+		return TC_ACT_UNSPEC;
+	}
+	__s8 dscp = EGRESS_DSCP;
+	CALI_DEBUG("setting dscp to %d", dscp);
+
+#ifdef IPVER6
+	// In IPv6, traffic class (8bits) equals to DSCP (6bits) + ECN (2bits). The 4 most significant bits of
+	// traffic class are stored in IPv6 priority field (4 bits), and the 4 least significant bits of it
+	// are stored in the 4 most significant bits of IPv6 flow_lbl[0] field. We must not change ECN bits here.
+	ip_hdr(ctx)->priority = (__u8) (dscp >> 2);
+	ip_hdr(ctx)->flow_lbl[0] = (__u8) (ip_hdr(ctx)->flow_lbl[0] & 0x3f) | (dscp << 6);
+#else
+	// In IPv4, DSCP (6bits) is located at the most significant bits of IPv4 TOS field.
+	// The 2 least significant bits are assigned to ECN and must not be touched.
+	ip_hdr(ctx)->tos = (__u8) ((ip_hdr(ctx)->tos & 0x03) | (dscp << 2));
+
+	__wsum ip_csum = bpf_csum_diff(0, 0, (__u32 *)ctx->ip_header, sizeof(struct iphdr), 0);
```

Comment:
don't you need to set the csum field to 0 first if you want to compute csum of the entire header? `This includes the Header Checksum field itself, which is set to zero during computation. `